### DISCUSSION TITLE
running pyvista on headless machine with Centos OS

### DISCUSSION THREAD

**Discussion by chen112p** [pyvista/pyvista#2615] [Q&A]
Hello, I understand this topic has been discussed few times. Wondering if anyone made it work on a headless virtual machine with Centos OS. All I am trying to do is to load existing vtk files and put them together as a gif animation.
I have installed Xvfb
![image](https://user-images.githubusercontent.com/39268233/138580180-c300392d-4aeb-4982-a47f-95e77d9d12ca.png)
and mesa-libGL. Also, created a shell script to config Xvfb
![image](https://user-images.githubusercontent.com/39268233/138580257-4a84fb73-8344-47ba-abd4-3c2baa8461ea.png)

But still when running the python script, it tells me 'could not find a decent visual'
![image](https://user-images.githubusercontent.com/39268233/138580302-ea41c459-f690-4dfb-9359-cb8751e77992.png)

I have checked that there is a xvfb process running in the background:
![image](https://user-images.githubusercontent.com/39268233/138580338-53792e81-a516-42d5-b6f7-53853b8c988d.png)

Can you tell what i missed and where i did wrong here?

Thanks!

Junting

_Originally posted by @chen112p in https://github.com/pyvista/pyvista-support/issues/190#issuecomment-950256838_

**Comment by redoules** [comment]
Hello,
Xvfg seems to be unmaintained. 
I've got some good results with XDummy.

First I made sure OpenGL is installed with xdummy

```
yum install mesa-libGL mesa-dri-drivers xorg-x11-drv-dummy xserver-xorg-video-dummy
```

I then created a config file for XDummy (dummy-1920x1080.conf)

```
Section "Monitor"
  Identifier "Monitor0"
  HorizSync 28.0-80.0
  VertRefresh 48.0-75.0
  Modeline "1920x1080_60.00" 172.80 1920 2040 2248 2576 1080 1081 1084 1118 -HSync +Vsync
EndSection

Section "Device"
  Identifier "Card0"
  Driver "dummy"
  VideoRam 256000
EndSection

Section "Screen"
  DefaultDepth 24
  Identifier "Screen0"
  Device "Card0"
  Monitor "Monitor0"
  SubSection "Display"
    Depth 24
    Modes "1920x1080_60.00"
  EndSubSection
EndSection
```


I then started Xdummy on the server 

```
sudo X :99 -config dummy-1920x1080.conf
```
Once the server started, make sure you export the environnement variables 

```
export DISPLAY=:99
export MESA_GL_VERSION_OVERRIDE=3.2
```

After than I managed to run a python program that was using VTK
```python headless.py```


Note that for pyvista, you might need to export some additionnal env variables

```
export PYVISTA_OFF_SCREEN=true
export PYVISTA_USE_IPYVTK=true
```


Useful links for reference :
https://stackoverflow.com/questions/8545291/opengl-glx-extension-not-supported
https://stackoverflow.com/questions/27407774/get-supported-glsl-versions
https://docs.mesa3d.org/envvars.html
https://github.com/gcaufield/MonkeyContainer/blob/master/dummy-1920x1080.conf
https://vtk.org/pipermail/vtkusers/2018-April/101556.html
